    As an introduction to natural language processing two texts will be obtained, preprocessed, tokenized, normalized and the most common tokens representing the texts will be examined.  These texts will then be subjected to the same process using bigrams and the top normalized tokenized bigrams will be examined.  The steps and decisions will be explained along the way.  

    The two texts are a Harry Potter and the Philosophers Stone by J. K. Rowling and Game of Thrones (A song of Ice and Fire, Book 1).  Both books are the first book in a series, and both are in the fantasy genre.  The differences between them are that one is for adult audiences, and one is for children.  The overall focus of the fantasy is also different, along with the time period and the use of fiction.  It is the goal to be able to see major differences in vocabulary representing the intended age range the books were designed for.  

    (1b) A text file containing the first Harry Potter book was obtained from Kaggle (https://www.kaggle.com/datasets/balabaskar/harry-potter-books-corpora-part-1-7).  It was read into python and the text was inspected to determine preprocessing steps.  An example section of the text is provided to show some of the preprocessing required.  There are many things that need to be preprocessed before the tokenization and normalization can be performed.  There are lines that are skipped (\n), page numbers, and some strange characters that appeared later in the text that required removal

In [3]:
# saving the path
hp_path = r'C:\Users\Morga\programsMG\NaturalLanguageProcessing\book1.txt'

# open the file
f = open(hp_path, 'r',encoding="utf8")

#reading the text saving it to a variable
hptext = f.read()

#displaying the section of the text
hptext[:1200]

'/ \n\n\n\n\nTHE BOY WHO LIVED \n\nMr. and Mrs. Dursley, of number four, Privet Drive, \nwere proud to say that they were perfectly normal, \nthank you very much. They were the last people you’d \nexpect to be involved in anything strange or \nmysterious, because they just didn’t hold with such \nnonsense. \n\nMr. Dursley was the director of a firm called \nGrunnings, which made drills. He was a big, beefy \nman with hardly any neck, although he did have a \nvery large mustache. Mrs. Dursley was thin and \nblonde and had nearly twice the usual amount of \nneck, which came in very useful as she spent so \nmuch of her time craning over garden fences, spying \non the neighbors. The Dursley s had a small son \ncalled Dudley and in their opinion there was no finer \nboy anywhere. \n\nThe Dursleys had everything they wanted, but they \nalso had a secret, and their greatest fear was that \nsomebody would discover it. They didn’t think they \ncould bear it if anyone found out about the Potters

    (2)Regular expressions patterns were created and saved using re.compile.  These patterns were placed inside of re.sub and the text was removed. 

In [84]:
import re

#Removing the \n
pattern = re.compile('\n')
hptext2=re.sub(pattern,'',hptext)

#removing Page | ### Harry Potter and the Philosophers Stone - J.K. Rowling
pattern2 = re.compile('Page \| [0-9][0-9]*[0-9]* Harry Potter and the Philosophers Stone - *J.K. Rowling')
hptext2=re.sub(pattern2,'',hptext2)

#removing this wierd k k k  and symbol k k k
pattern3 = re.compile('•k k k|k k k')
hptext2=re.sub(pattern3,'',hptext2)

    There remain a few items at the beginning and end of the text that require removal.  These were removed directly after tokenization.  Using nltk word tokenizer the text was broken in word tokens.

In [85]:
#tokenization
import nltk
hptokens = nltk.word_tokenize(hptext2)

    The beginning and end of the token list was examined to see the tokens that were created that are not part of the actual book but are from the text file.

In [86]:
#showing the beginging and ending
print(len(hptokens))
print(hptokens[:10])
print(hptokens[101200:])

101215
['/', 'THE', 'BOY', 'WHO', 'LIVED', 'Mr.', 'and', 'Mrs.', 'Dursley', ',']
['’', 'm', 'going', 'to', 'have', 'a', 'lot', 'of', 'fun', 'with', 'Dudley', 'this', 'summer', '...', '”']


    These tokens are manually deleted.

In [87]:
#the last token is not from the book needs to be removed
del hptokens[101214]
print(hptokens[101200:])

['’', 'm', 'going', 'to', 'have', 'a', 'lot', 'of', 'fun', 'with', 'Dudley', 'this', 'summer', '...']


In [88]:
#The first token is not from the book needs to be removed
del hptokens[0]
print(hptokens[:10])

['THE', 'BOY', 'WHO', 'LIVED', 'Mr.', 'and', 'Mrs.', 'Dursley', ',', 'of']


    (1b)A text file containing the first Game of Thrones book was obtained from Kaggle (https://www.kaggle.com/datasets/khulasasndh/game-of-thrones-books).  Again this data was read into python and the text was inspected to determine preprocessing steps.  An example section of the text is provided to show some of the preprocessing required.  There are many things that need to be preprocessed before the tokenization and normalization can be performed.  There are lines that are skipped (\n) and page numbers.  It was also found that there were numbers replacing letters in the text itself.  There are o represented by 0.  I and L represented by 1s and more.


In [89]:
# saving the path
got_path = r'C:\Users\Morga\programsMG\NaturalLanguageProcessing\001ssb.txt'

# open the file
f = open(got_path, 'r',encoding="utf8")

#reading the text and saving it to a variable
gottext = f.read()

#displaying the section of the text
gottext[:1000]

'A Game Of Thrones \nBook One of A Song of Ice and Fire \nBy George R. R. Martin \nPROLOGUE \n"We should start back," Gared urged as the woods began to grow dark around them. "The wildlings are \ndead." \n"Do the dead frighten you?" Ser Waymar Royce asked with just the hint of a smile. \nGared did not rise to the bait. He was an old man, past fifty, and he had seen the lordlings come and go. \n"Dead is dead," he said. "We have no business with the dead." \n"Are they dead?" Royce asked softly. "What proof have we?" \n"Will saw them," Gared said. "If he says they are dead, that\'s proof enough for me." \nWill had known they would drag him into the quarrel sooner or later. He wished it had been later rather \nthan sooner. "My mother told me that dead men sing no songs," he put in. \n"My wet nurse said the same thing, Will," Royce replied. "Never believe anything you hear at a woman\'s \ntit. There are things to be learned even from the dead." His voice echoed, too loud in the twilit fores

    (2)Regular expressions patterns were created and saved using re.compile.  These patterns were placed inside of re.sub and the text was removed. There was a large amount of preprocessing required for this book.

In [90]:
#Removing the \n
pattern = re.compile('\n')
gottext2=re.sub(pattern,'',gottext)


## Removing page numbers and aurhtor from the pages along with some random numbers typed it the text
pattern2 = re.compile('[Pp]age [0-9][0-9]*[0-9]*')
gottext2=re.sub(pattern2,'',gottext2)

pattern3 = re.compile('1fm')
gottext2=re.sub(pattern3,"I'm",gottext2)

pattern4 = re.compile('m1ord')
gottext2=re.sub(pattern4,"mlord",gottext2)

pattern5 = re.compile('11 \*')
gottext2=re.sub(pattern5,"I'll",gottext2)

pattern6 = re.compile('Aerys 11 Targaryen')
gottext2=re.sub(pattern6,"Aerys II Targaryen",gottext2)

pattern7 = re.compile('Aegon 111')
gottext2=re.sub(pattern7,"Aegon III",gottext2)

pattern8 = re.compile('11IPm')
gottext2=re.sub(pattern8,"I'm",gottext2)

pattern9 = re.compile('CAME OF THRONES 119')
gottext2=re.sub(pattern9,"",gottext2)

pattern10 = re.compile('0')
gottext2=re.sub(pattern10,"o",gottext2)

pattern11 = re.compile('A GAM[EL],* OF THRON[LE]S [0-9][0-9]*[0-9]*')
gottext2=re.sub(pattern11,"of",gottext2)

pattern12 = re.compile('[0-9][0-9]*[0-9]* GLORG[LE] R.R. MARTIN')
gottext2=re.sub(pattern12,"of",gottext2)

pattern13 = re.compile('402 GEORGE R.R. MARUN')
gottext2=re.sub(pattern13,"",gottext2)

pattern14 = re.compile('11')
gottext2=re.sub(pattern14,"",gottext2)

pattern15 = re.compile('[0-9][0-9]*[0-9]* GEORGL R.R. MARTIN')
gottext2=re.sub(pattern15,"",gottext2)

pattern16 = re.compile('1')
gottext2=re.sub(pattern16,"I",gottext2)

pattern17 = re.compile('[0-9][0-9]*[0-9]* GEORGE R.R. MARTM')
gottext2=re.sub(pattern17,"",gottext2)

pattern18 = re.compile('1oo GLORGE R.R. MARTIN')
gottext2=re.sub(pattern18,"",gottext2)

pattern19 = re.compile('[0-9][0-9]*[0-9]* GEORGE RA. MARTIN')
gottext2=re.sub(pattern19,"",gottext2)

pattern20 = re.compile('[0-9][0-9]*[0-9]* GEORGE R.R. MARTIN')
gottext2=re.sub(pattern20,"",gottext2)

pattern21 = re.compile('[0-9][0-9]*[0-9]* GLORGL R.R. MARTIN')
gottext2=re.sub(pattern21,"",gottext2)

pattern22 = re.compile('A [CG]A[NM]4*[EL] OF THRON[EF]S [0-9][0-9]*[0-9]')
gottext2=re.sub(pattern22,"",gottext2)

pattern23 = re.compile('9\' ". . .')
gottext2=re.sub(pattern23,"",gottext2)

pattern24 = re.compile('[0-9][0-9]*[0-9]* GEORGE R.R. MARUN')
gottext2=re.sub(pattern24,"",gottext2)

pattern25 = re.compile('A GAME OF TFIRONES 535')
gottext2=re.sub(pattern25,"",gottext2)

pattern26 = re.compile('[0-9][0-9]*[0-9]* GLORGL R.R. MARTIN')
gottext2=re.sub(pattern26,"",gottext2)

pattern27 = re.compile('26o GLORGL R.R. MARTIN')
gottext2=re.sub(pattern27,"",gottext2)

pattern28 = re.compile('ofo[0-9]')
gottext2=re.sub(pattern28,"",gottext2)

pattern29 = re.compile('m7ord')
gottext2=re.sub(pattern29,"mlord",gottext2)

pattern30 = re.compile('4o ')
gottext2=re.sub(pattern30,"",gottext2)

pattern31 = re.compile('Red4yne')
gottext2=re.sub(pattern31,"Redyne",gottext2)

pattern32 = re.compile('65o GLORGL R.R. MARTIN')
gottext2=re.sub(pattern32,"",gottext2)

pattern33 = re.compile('Red4yne')
gottext2=re.sub(pattern33,"Redyne",gottext2)

pattern34 = re.compile("A GAME OF 'FHRONES 679")
gottext2=re.sub(pattern34,"",gottext2)

pattern35 = re.compile(" I 9")
gottext2=re.sub(pattern35,"",gottext2)

pattern36=re.compile("\'\'")
gottext2=re.sub(pattern36,"",gottext2)

    Using nltk word tokenizer the text was broken in word tokens.  There remains one pattern that was not able to be deleted by using Regular expressions and was removed directly after tokenization.

In [91]:
#tokenization
gottokens = nltk.word_tokenize(gottext2)

In [92]:
# creating a loop to manually remove an unwanted token.
fixedgottokesn =[]
for token in gottokens:
    if token == "''":
        continue
    else:
        fixedgottokesn.append(token)

    To visualize and understand how each step is reducing the Frequency distribution and total vocabulary a function is created for easier access to this information.  It simply takes the tokenized text as input and prints out the FreqDist function provided by nltk then using a loop displays the most common 15 tokens and the number of times they appear in the text.

In [93]:
from nltk import FreqDist

#creating a function to display
#frequeny distribution and total vocab count
#also prints most common 15 tolkens
def vocab(list):
    ndist=FreqDist(list)
    nitems = ndist.most_common(15)
    print(FreqDist(list))
    for item in nitems:
        print(item)

In [94]:
# Looking at the original tokens created
print('hp:')
print(vocab(hptokens))
print('got:')
print(vocab(fixedgottokesn))

hp:
<FreqDist with 7129 samples and 101213 outcomes>
(',', 5658)
('.', 4642)
('the', 3312)
('’', 3111)
('“', 2437)
('”', 2413)
('to', 1844)
('and', 1808)
('a', 1579)
('Harry', 1318)
('of', 1239)
('he', 1208)
('was', 1179)
('s', 1002)
('in', 931)
None
got:
<FreqDist with 13187 samples and 353775 outcomes>
('.', 25060)
(',', 21040)
('the', 15126)
('and', 8419)
('``', 8167)
('to', 6490)
('a', 5853)
('of', 5825)
('his', 4597)
('was', 3890)
('he', 3428)
('her', 3416)
('I', 3219)
('in', 3092)
('had', 2855)
None


	(2a)To reduce the vocabulary and combine words that start a sentence and are used in a sentence (non-proper nouns) all the text will be converted to lower case.  This was accomplished by creating a function that takes each token from the list and uses .lower() on the token saves it as a temporary variable then adds it to a temporary list and returns the temporary list.  This function was used on each of the tokenized books and saved as a new list.  In some cases, the upper-case letters could give valuable information, they specify proper nouns and the start of sentences.  The word used to begin each sentence is not important to this analysis and the proper nouns are already known.

In [95]:
#creating a function to take a list 
#make each token lowercase
#place the lowercase tokens into a list
def lc(list):
    lclist =[]
    for token in list:
        lcw = token.lower()
        lclist.append(lcw) 
    return lclist

In [96]:
#using the lowercase function
# creating new tokenized list
#only lowercase characters remain
hptokens_lc=lc(hptokens)
gottokens_lc=lc(fixedgottokesn)

In [97]:
print('hp:')
print(vocab(hptokens_lc))
print('got:')
print(vocab(gottokens_lc))

hp:
<FreqDist with 6295 samples and 101213 outcomes>
(',', 5658)
('.', 4642)
('the', 3626)
('’', 3111)
('“', 2437)
('”', 2413)
('and', 1920)
('to', 1854)
('he', 1756)
('a', 1684)
('harry', 1319)
('of', 1256)
('was', 1186)
('it', 1168)
('s', 1003)
None
got:
<FreqDist with 11864 samples and 353775 outcomes>
('.', 25060)
(',', 21040)
('the', 17689)
('and', 8834)
('``', 8167)
('to', 6546)
('a', 6384)
('of', 5838)
('he', 5402)
('his', 5073)
('was', 3915)
('her', 3642)
('you', 3472)
('in', 3229)
('i', 3220)
None


	(2a)As is shown above the top 15 tokens of each book are comprised of mostly stop words and punctuation.  The objective is to compare vocabulary used between two different fantasy books one for adults and one for younger audiences, the punctuation must be removed.  Using the punctuation list imported from string and additional punctuation marks found not to be included were used in a function.  The function takes each token and checks if the token is not in the punctuation list created.  If the token is not on the punctation list, that token is placed in a new list.  The new list is returned. This function is run on the lowercase only tokenized list and saved as a new list.  The added punctuation was important because after initially running the “normal” punctuation list the added ones were still found to be in the most common 15.  This is taking up a place of some frequent vocabulary of interest.

In [98]:
#creating a function to remove punctuation from a list
from string import punctuation
punctuation = punctuation + '’”“`''`—'
def remove_punc(tokens):
    return [t for t in tokens if t not in punctuation]

In [99]:
hptokens_lcp=remove_punc(hptokens_lc)
gottokens_lcp=remove_punc(gottokens_lc)

In [100]:
print('hp:')
print(vocab(hptokens_lcp))
print('got:')
print(vocab(gottokens_lcp))

hp:
<FreqDist with 6281 samples and 80580 outcomes>
('the', 3626)
('and', 1920)
('to', 1854)
('he', 1756)
('a', 1684)
('harry', 1319)
('of', 1256)
('was', 1186)
('it', 1168)
('s', 1003)
('you', 992)
('in', 963)
('his', 936)
('i', 921)
('t', 833)
None
got:
<FreqDist with 11851 samples and 296750 outcomes>
('the', 17689)
('and', 8834)
('to', 6546)
('a', 6384)
('of', 5838)
('he', 5402)
('his', 5073)
('was', 3915)
('her', 3642)
('you', 3472)
('in', 3229)
('i', 3220)
('it', 3058)
('had', 2868)
('she', 2865)
None


    (2a)(2d)As seen above the topmost common tokens are mostly comprised of stop words.  The most common tokens from both texts are very similar and will not yield good results since the vocabulary is to be compared.  To find the vocabulary that is used most often, but that gives actual meaning and understanding into the audiences age that the books are intended for, the stop words must be removed.  A list of stop words was acquired from nltk.  Since the tokenization process that was used created ‘s and n’t in separate tokens, they were added to the stop list.  The list of lowercase tokens that had punctuation removed were run through a loop each token was compared to the stop list.  If the token did not appear in the stop list, it was added into a new list.  This list no longer contained any of the tokens that appear on the stop list created.  Once this was preformed initially both lists had said as either the most common or second most common token.  Since this is a very common word in fiction books and added little information gain it was also added to the stop list. 

In [101]:
nltkstopwords = nltk.corpus.stopwords.words('english')
stopwords = nltkstopwords +  ["'s","n't","said"]

In [102]:

hp_stop=[]
for token in hptokens_lcp:
    if token not in stopwords:
        hp_stop.append(token)
    else:
        continue
        
got_stop=[]
for token in gottokens_lcp:
    if token not in stopwords:
        got_stop.append(token)
    else:
        continue

In [103]:
print('hp:')
vocab(hp_stop)
print('got:')
vocab(got_stop)

hp:
<FreqDist with 6131 samples and 40231 outcomes>
('harry', 1319)
('ron', 425)
('hagrid', 370)
('hermione', 267)
('back', 255)
('one', 251)
('...', 238)
('got', 206)
('could', 198)
('get', 194)
('like', 192)
('know', 181)
('see', 178)
('professor', 177)
('looked', 169)
got:
<FreqDist with 11717 samples and 148838 outcomes>
('lord', 1301)
('would', 1156)
('ser', 973)
('could', 840)
('jon', 830)
('one', 792)
('ned', 785)
('man', 775)
('king', 736)
('back', 691)
('father', 635)
('tyrion', 615)
('like', 597)
('hand', 584)
('men', 575)


	(2a)To reduce the vocabulary and combine words such as different and differently the porter stemmer was used to break the tokens down to its root.  This gives the chance to move a more commonly used word that has different suffix/prefix up in frequency. 

In [104]:
## Porter stemming
from nltk.stem.porter import PorterStemmer
ps = PorterStemmer()

In [105]:
hp_stem =[]
for token in hp_stop:
    x=ps.stem(token)
    hp_stem.append(x)
    
got_stem =[]
for token in got_stop:
    x=ps.stem(token)
    got_stem.append(x)

In [106]:
print('hp:')
print(vocab(hp_stem))
print('got:')
print(vocab(got_stem))

hp:
<FreqDist with 4557 samples and 40231 outcomes>
('harri', 1319)
('ron', 425)
('look', 404)
('hagrid', 370)
('hermion', 267)
('go', 263)
('one', 262)
('back', 261)
('...', 238)
('get', 235)
('know', 212)
('like', 211)
('got', 206)
('could', 198)
('see', 183)
None
got:
<FreqDist with 7725 samples and 148838 outcomes>
('lord', 1438)
('would', 1156)
('ser', 975)
('could', 840)
('hand', 833)
('jon', 830)
('one', 823)
('look', 814)
('ned', 785)
('man', 781)
('king', 775)
('back', 719)
('like', 715)
('father', 651)
('eye', 630)
None


	Below are the top 50 most common tokens from each text.  This will be used with the bigrams to examine the vocabulary and compare them to see the difference between fantasy books written for particular age ranges.

In [107]:
fdisthp = FreqDist(hp_stem)
hptopkeys = fdisthp.most_common(50)
for pair in hptopkeys:
    print(pair)

('harri', 1319)
('ron', 425)
('look', 404)
('hagrid', 370)
('hermion', 267)
('go', 263)
('one', 262)
('back', 261)
('...', 238)
('get', 235)
('know', 212)
('like', 211)
('got', 206)
('could', 198)
('see', 183)
('professor', 177)
('tri', 170)
('snape', 168)
('dumbledor', 158)
('think', 146)
('around', 141)
('dudley', 137)
('want', 135)
('time', 134)
('say', 129)
('come', 129)
('someth', 128)
('never', 126)
('malfoy', 126)
('uncl', 122)
('head', 121)
('door', 121)
('even', 120)
('eye', 120)
('right', 120)
('yeh', 119)
('nevil', 116)
('turn', 115)
('vernon', 115)
('well', 114)
('hand', 114)
('quirrel', 113)
('face', 111)
('would', 108)
('first', 108)
('potter', 107)
('dursley', 106)
('thing', 103)
('way', 101)
('ask', 101)


In [108]:
fdistgot = FreqDist(got_stop)
gottopkeys = fdistgot.most_common(50)
for pair in gottopkeys:
    print(pair)

('lord', 1301)
('would', 1156)
('ser', 973)
('could', 840)
('jon', 830)
('one', 792)
('ned', 785)
('man', 775)
('king', 736)
('back', 691)
('father', 635)
('tyrion', 615)
('like', 597)
('hand', 584)
('men', 575)
('eyes', 561)
('bran', 546)
('told', 502)
('see', 498)
('catelyn', 491)
('arya', 459)
('even', 447)
('face', 443)
('boy', 439)
('brother', 437)
('know', 423)
('sansa', 422)
('dany', 413)
('robb', 409)
('old', 405)
('looked', 405)
('never', 404)
('robert', 403)
('well', 397)
('time', 393)
('stark', 393)
('black', 390)
('long', 389)
('yet', 378)
('head', 374)
('away', 369)
('made', 361)
('night', 354)
('maester', 344)
('lannister', 344)
('lady', 343)
('come', 337)
('took', 337)
('thought', 336)
('sword', 335)


	Nltk’s bigrams was used on the list of tokens containing on lower case characters and placed in a list. The bigram assocmeasures() was saved into a variable and that variable was saved as finder and used to create a score for the bigrams by frequency.  As expected much normalization needs to occur before any information is gained from the bigrams.

In [109]:
hpbigrams = list(nltk.bigrams(hptokens_lc))
gotbigrams = list(nltk.bigrams(gottokens_lc))

In [111]:
# setup for bigrams and bigram measures
from nltk.collocations import *
bigram_measures = nltk.collocations.BigramAssocMeasures()

In [135]:
# create the bigram finder and score the bigrams by frequency
finder = BigramCollocationFinder.from_words(hptokens_lc)
scored = finder.score_ngrams(bigram_measures.raw_freq)

In [136]:
# create the bigram finder and score the bigrams by frequency
finder = BigramCollocationFinder.from_words(hptokens_lc)
scored = finder.score_ngrams(bigram_measures.raw_freq)
# scores are sorted in decreasing frequency
for bscore in scored[:15]:
    print (bscore)

(('.', '“'), 0.015343878750753362)
(('’', 's'), 0.009870273581456928)
(('’', 't'), 0.008230168061415036)
((',', '”'), 0.0070050289982512125)
(('”', '“'), 0.00596761285605604)
(('”', 'said'), 0.0055526463991779716)
(('?', '”'), 0.005167320403505479)
((',', 'and'), 0.004021222570223193)
(('.', 'he'), 0.0038532599567249266)
((',', 'but'), 0.0029047651981464833)
(('!', '”'), 0.002894885044411291)
(('of', 'the'), 0.002825723968264946)
(('i', '’'), 0.0026972819697074486)
(('’', 'd'), 0.0026676415085018722)
(('in', 'the'), 0.0026577613547666802)


In [137]:
# create the bigram finder and score the bigrams by frequency
finder2 = BigramCollocationFinder.from_words(gottokens_lc)
scored2 = finder2.score_ngrams(bigram_measures.raw_freq)
# scores are sorted in decreasing frequency
for bscore in scored2[:15]:
    print (bscore)

(('.', '``'), 0.019633948130874143)
((',', 'and'), 0.007725249099003604)
(('.', 'the'), 0.00555437778248887)
(('.', 'he'), 0.004692247897675076)
(('.', '.'), 0.004655501377994488)
(('of', 'the'), 0.004293689491908699)
((',', 'but'), 0.0032393470426118295)
((',', 'the'), 0.003160200692530563)
(('in', 'the'), 0.00312062751748993)
(('``', 'i'), 0.002934068263726945)
((',', 'he'), 0.002654229383082468)
(('.', 'she'), 0.0023291640166772667)
(('said', '.'), 0.00218500459331496)
(('and', 'the'), 0.0020295385485124724)
(('to', 'the'), 0.002006925305632111)


	(2a)Since the above bigrams are not giving much information in terms of vocabulary punctuation must first be removed.  In this case to remove any token that contains non-alphabetical characters the alpha_filter is used.  This will allow only vocabulary to remain which is what is required for the comparison.

In [138]:
def alpha_filter(w):
  # pattern to match word of non-alphabetical characters
    pattern = re.compile('^[^a-z]+$')
    if (pattern.match(w)):
        return True
    else:
        return False

In [139]:
# apply a filter to remove non-alphabetical tokens from the emma bigram finder
finder.apply_word_filter(alpha_filter)
scored = finder.score_ngrams(bigram_measures.raw_freq)
for bscore in scored[:15]:
    print(bscore)

(('of', 'the'), 0.002825723968264946)
(('in', 'the'), 0.0026577613547666802)
(('on', 'the'), 0.0021143528993311134)
(('it', 'was'), 0.0020353116694495767)
(('he', 'was'), 0.0019266299783624633)
(('to', 'the'), 0.0016993864424530445)
(('out', 'of'), 0.0014425024453380494)
(('at', 'the'), 0.0013931016766620889)
(('said', 'harry'), 0.0013338207542509361)
(('he', 'had'), 0.0011362176795470939)
(('said', 'ron'), 0.001076936757135941)
(('to', 'be'), 0.001067056603400749)
(('uncle', 'vernon'), 0.001067056603400749)
(('he', 'said'), 0.0010472962959303647)
(('in', 'a'), 0.0010374161421951725)


In [140]:
# apply a filter to remove non-alphabetical tokens from the emma bigram finder
finder2.apply_word_filter(alpha_filter)
scored2 = finder2.score_ngrams(bigram_measures.raw_freq)
for bscore in scored2[:15]:
    print(bscore)

(('of', 'the'), 0.004293689491908699)
(('in', 'the'), 0.00312062751748993)
(('and', 'the'), 0.0020295385485124724)
(('to', 'the'), 0.002006925305632111)
(('it', 'was'), 0.0016507667302664124)
(('on', 'the'), 0.0015631404141050102)
(('he', 'was'), 0.0015094339622641509)
(('from', 'the'), 0.0013313546745813017)
(('the', 'king'), 0.0012295950816196735)
(('he', 'had'), 0.0011815419404989046)
(('at', 'the'), 0.0011504487315384072)
(('had', 'been'), 0.0011023955904176383)
(('of', 'his'), 0.0011023955904176383)
(('with', 'a'), 0.001062822415377005)
(('he', 'said'), 0.0009751960992156032)


	(2a)(2d)As seen above the bigrams with the highest scores from both texts are comprised of mostly stop words.  Using a lambda function the stopwords list created before was used to remove stop words.  This portion also removes any bigrams were the frequency is less than 5 in the texts.

In [141]:
# apply a filter to remove stop words
finder.apply_freq_filter(5)
finder.apply_word_filter(lambda w: w in stopwords)
scored = finder.score_ngrams(bigram_measures.raw_freq)
for bscore in scored[:50]:
    print(bscore)

(('uncle', 'vernon'), 0.001067056603400749)
(('professor', 'mcgonagall'), 0.0008990939899024829)
(('aunt', 'petunia'), 0.0005137679942299903)
(('mr.', 'dursley'), 0.0002964046120557636)
(('harry', 'potter'), 0.0002667641508501872)
(('harry', 'could'), 0.00021736338217422663)
(('harry', 'looked'), 0.00021736338217422663)
(('looked', 'like'), 0.0002074832284390345)
(('professor', 'dumbledore'), 0.0002074832284390345)
(('common', 'room'), 0.0001976030747038424)
(('could', 'see'), 0.0001976030747038424)
(('harry', 'felt'), 0.0001976030747038424)
(('first', 'years'), 0.00018772292096865028)
(('mrs.', 'dursley'), 0.00018772292096865028)
(('professor', 'quirrell'), 0.00018772292096865028)
(('hermione', 'granger'), 0.00016796261349826604)
(('harry', 'asked'), 0.00015808245976307392)
(('harry', 'thought'), 0.00015808245976307392)
(('mr.', 'ollivander'), 0.00015808245976307392)
(('privet', 'drive'), 0.00015808245976307392)
(('great', 'hall'), 0.0001482023060278818)
(('nimbus', 'two'), 0.00014820

In [142]:
# apply a filter to remove stop words
finder2.apply_freq_filter(5)
finder2.apply_word_filter(lambda w: w in stopwords)
scored2 = finder2.score_ngrams(bigram_measures.raw_freq)
for bscore in scored2[:50]:
    print(bscore)

(('ser', 'jorah'), 0.0004353049254469649)
(('lord', 'tywin'), 0.00035333192000565333)
(('maester', 'luwin'), 0.0003335453324853367)
(('ser', 'rodrik'), 0.0003194120556851106)
(('khal', 'drogo'), 0.0003081054342449297)
(('lord', 'eddard'), 0.0003024521235248392)
(('could', 'see'), 0.00026005229312416083)
(('lord', 'father'), 0.00023743905024379902)
(('tyrion', 'lannister'), 0.0002261324288036181)
(('septa', 'mordane'), 0.00021199915200339198)
(('eddard', 'stark'), 0.00018090594304289448)
(('ser', 'barristan'), 0.00018090594304289448)
(('grand', 'maester'), 0.00017807928768284926)
(('seven', 'kingdoms'), 0.00017807928768284926)
(('lord', 'commander'), 0.00017525263232280404)
(('lord', 'renly'), 0.00017242597696275882)
(('could', 'feel'), 0.00016394601088262314)
(('ser', 'alliser'), 0.00016394601088262314)
(('old', 'man'), 0.00015546604480248745)
(('maester', 'pycelle'), 0.00015263938944244223)
(('would', 'never'), 0.000149812734082397)
(('maz', 'duur'), 0.0001469860787223518)
(('jon', 'a

    (2b) When looking at the lists of most frequent tokens and the most frequent bigrams they are a strong representation of the text themselves.  It would be possible to remove Proper Nouns (character names) to increase the vocabulary gained.  Although this process would be quite extensive as there are many included, and as they were removed more would most likely take their place knowing these books have a very high number of characters.

    The bigrams are run with pointwise mutual information and the top 50 are displayed.  This provides information that these words express a unique concept when used together and may give a better expression of intent

In [143]:
### pointwise mutual information
finder.apply_freq_filter(5)
scored = finder.score_ngrams(bigram_measures.pmi)
for bscore in scored[:50]:
    print (bscore)

(('boa', 'constrictor'), 14.305106984117373)
(('goal', 'posts'), 13.819680156947133)
(('diagon', 'alley'), 13.627035079004735)
(('flavor', 'beans'), 13.457110077562422)
(('hospital', 'wing'), 13.364000673170938)
(('smelting', 'stick'), 13.305106984117373)
(('shooting', 'stars'), 13.194075671728626)
(('bloody', 'baron'), 12.872147576841268)
(('lee', 'jordan'), 12.845675365480078)
(('marcus', 'flint'), 12.819680156947129)
(('leaky', 'cauldron'), 12.457110077562422)
(('privet', 'drive'), 12.305106984117376)
(('marble', 'staircase'), 12.248523455751005)
(('chocolate', 'frogs'), 12.211997579725892)
(('seamus', 'finnigan'), 12.082714562780925)
(('fat', 'lady'), 11.912789561338617)
(('portrait', 'hole'), 11.819680156947133)
(('vault', 'seven'), 11.64975515550482)
(('three-headed', 'dog'), 11.546661662540716)
(('nicolas', 'flamel'), 11.479136383892424)
(('madam', 'hooch'), 11.457110077562422)
(('madam', 'malkin'), 11.457110077562422)
(('madam', 'pomfrey'), 11.457110077562422)
(('nearly', 'head

In [144]:
### pointwise mutual information
finder2.apply_freq_filter(5)
scored2 = finder2.score_ngrams(bigram_measures.pmi)
for bscore in scored2[:50]:
    print (bscore)

(('yi', 'ti'), 16.110544481138117)
(('helman', 'tallhart'), 15.84751007530432)
(('mance', 'rayder'), 15.110544481138117)
(('masha', 'heddle'), 15.110544481138117)
(('aron', 'santagar'), 14.99951316874937)
(('tobho', 'mott'), 14.947045748855235)
(('lyn', 'corbray'), 14.940619479695805)
(('arthur', 'dayne'), 14.89997749519846)
(('dosh', 'khaleen'), 14.625117653967877)
(('balon', 'swann'), 14.625117653967875)
(('donal', 'noye'), 14.625117653967873)
(('willis', 'wode'), 14.525581980416959)
(('karyl', 'vance'), 14.414550668028216)
(('huny', 'huny'), 14.373578886971908)
(('hallis', 'mollen'), 14.345009734775138)
(('galbart', 'glover'), 14.236075363221975)
(('jason', 'mallister'), 14.21745968505463)
(('two-handed', 'greatsword'), 14.142965958830494)
(('roose', 'bolton'), 14.10208290238992)
(('bowen', 'marsh'), 14.040155153246719)
(('flea', 'bottom'), 13.97304095738818)
(('moat', 'cailin'), 13.908910619968466)
(('janos', 'slynt'), 13.870230151804405)
(('marq', 'piper'), 13.815088597611943)
(('

	(2c) The PMI removed some of the names of characters that appeared in the bigrams.  This is because they are very common characters, and their names are not first and last each time they are used.  Although it does give new character names, that are only used in full as their replacements.  The PMI does give a good feel for the books themselves.  Capturing completely made-up names of people and places in the Harry Potter books any fan would be familiar with such as Diagon Alley, Chocolate Frogs, and three-headed dog.  While in the Game of Thrones book it showcases darker terms and an air of medieval times such as two-handed greatsword, flea bottom, and cuts deeper.  

In [145]:
print(FreqDist(hp_stem))
print(FreqDist(got_stem))

<FreqDist with 4557 samples and 40231 outcomes>
<FreqDist with 7725 samples and 148838 outcomes>


    	As a review the goal is to see difference in the top tokens that would distinguish one fantasy book as being for adults (Game of Thrones) (GoT) and one for children (Harry Potter) (HP).  When looking at the top tokens after normalization we can see names in both, around the same number (13 in HP and 12 in GoT) the other words are telling of the books.  HP simple non descriptive words, (go, could, something) and Professor while GoT has some simple non descriptive words (looked, like) but also sword, king, lord, lady and maester a term from the book (a holy man).     “ …:” was left in as a token because it used 238 times in Harry Potter, this signifies almost overuse possibly because it is designed for children.  We see much more distinctive words from GoT which was intended for adults.  This is most likely caused by a variety of synonyms and a larger overall unique vocabulary.  This can also be seen in the Frequncy distribution from HP being 4557 while in GoT being 7725.  

    In the bigrams 11 of the 50 are Harry something. This is Harry felt, Harry asked, Harry looked, etc..  It is telling of the book being all about Harry and his perspective of the world.  While we see in GoT there are a lot more names. There are also more complex world building bigrams like titles (lord commander, grand measter).  HP has places, yet they describe smaller settings such as, common room, great hall, private drive, and one playful “leaky cauldron”. while GoT has many places that are larger in scale, Seven Kingdoms, Casterly Rock, Grey Wind, and Red Keep. This shows the size of the world being created in the text is very different.  The smaller world being created is from the child’s fantasy book while the larger is of the adult fantasy book.

    In the Pointwise mutual information, there are a lot of side character names that are always mentioned in full.  Then in HP there is portion of the story telling that is mystical, smelting stick, flavor beans, goal posts, chocolate frogs, forbidden forest, unicorn blood, dark arts.  We see one example that is comical for children but not so funny for adults “fat lady”.  This is a intricate world that has been created but in scales it is smaller.  When we look at the GoT PMI there are many more side characters.  So many that it is at times hard to see that they are proper nouns.  The next big separation between texts breaks through and that is exampled in yi ti, tobho mott, vaes Dothrak, and maz duur.  These seem like misspelling or possibly something went wrong in python coding.  But a small bit of understanding can shed light that this is Dothraki a language created for this series.  The PMI picks these words out because they don’t appear alone.  This sheds light into an aspect of fantasy that isn’t very often found in children’s books.  The writer creates another langue specifically for the book.

	Overall, the words presented in top words, bigrams, and the PMI bigrams, point to the size of the fictious world that was created in these texts.  While both are complex and other worldly.  The overall tokens apprehended point to the Game of Thrones world being much larger in scale and shown with the PMI in shear number of characters.  This largeness is more suited for an advanced reader.  The elements particular to the books pointed out by the PMI are fun and very understandable for Harry Potter, while in the Game of Thrones the PMI finds a language that was created for the books and a staggering number of side characters.
    
    The complexity difference in the findings showcases one of the largest differences in fantasy books written for adults vs children, the shear size of the worlds being created.
